### DISCUSSION TITLE
NaNs

### DISCUSSION THREAD

**Discussion by OlafvdSpek** [nlohmann/json#2914] [Q&A]
AFAIK NaNs get written as nulls and nulls throw when being read, so I wrote code like this:

```
j.at("x").is_null() ? std::numeric_limits<float>::quiet_NaN() : j.at("x").get<float>();
```

Is there a simpler way to write this, without creating my own wrapper function? Something like `j.at("x").value_or(default)` / `j.at("x").value_or(std::numeric_limits<float>::quiet_NaN())`?

It also means I can't use `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE`, is there a way to maybe define a 'global' conversion function for this, that could map null to nan?

**Comment by nlohmann** [comment]
No, there is no other way.

**Reply by OlafvdSpek** [reply]
Wouldn't a value_or(default) method be handy?

**Reply by nlohmann** [reply]
Like https://json.nlohmann.me/api/basic_json/value/ ?

**Reply by OlafvdSpek** [reply]
Yes, though that one doesn't take care of null values AFAIK, it only takes care of non-existing keys.

**Reply by nlohmann** [reply]
Indeed. But a conversion from `null` to NaN is really domain-specific. I'm not sure there is a generic solution that fits all clients.

**Reply by OlafvdSpek** [reply]
The NaN is domain-specific indeed, but that'd be passed in as an argument. Using a default when a value is null doesn't seem domain-specific and would be broadly useful IMO.

**Reply by gregmarr** [reply]
It may be broadly useful, but it's a breaking change in behavior at this point.
